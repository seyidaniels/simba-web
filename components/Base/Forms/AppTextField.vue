<template>
  <div>
    <v-text-field
      outlined
      :type="fieldType"
      hide-details="auto"
      v-bind="{ ...$props, ...$attrs }"
      v-on="$listeners"
      @input="handleInput"
    >
      <template v-for="(index, name) in $slots" #[name]>
        <slot :name="name" />
      </template>
      <template #append>
        <template v-if="$attrs.type === 'password' && visibilityToggle">
          <div class="password-toggle" @click="handleVisibilityToggle">
            <p v-if="fieldType === 'password'">Show</p>
            <p v-else>Hide</p>
          </div>
        </template>
        <template v-else>
          <template
            v-if="
              $attrs.error ||
              ($attrs['error-messages'] && $attrs['error-messages'].length)
            "
          >
            <v-icon small color="error">fa-info-circle</v-icon>
          </template>
        </template>
      </template>
    </v-text-field>
    <v-slide-y-transition>
      <div
        v-if="emailAutoComplete && showEmailAutoComplete"
        class="auto-domains"
      >
        <p
          v-for="(domain, index) in autoCompleteDomains"
          :key="index"
          v-ripple="{ class: 'neutral--text' }"
          @click="completeEmail(domain)"
        >
          {{ domain }}
        </p>
      </div>
    </v-slide-y-transition>
  </div>
</template>

<script>
export default {
  name: 'AppTextField',
  props: {
    /*
     * @prop height: Sets the height of the input
     * */
    height: {
      type: Number,
      default: 50,
    },
    visibilityToggle: {
      type: Boolean,
      default: false,
    },
    emailAutoComplete: {
      type: Boolean,
      default: false,
    },
    autoCompleteDomains: {
      type: Array,
      default: () => ['gmail.com', 'yahoo.com', 'hotmail.com'],
    },
  },
  data() {
    return {
      value: '',
      fieldType: 'text',
      showEmailAutoComplete: false,
    }
  },
  mounted() {
    this.fieldType = this.$attrs.type
  },
  methods: {
    handleVisibilityToggle() {
      this.fieldType === 'password'
        ? (this.fieldType = 'text')
        : (this.fieldType = 'password')
    },
    handleInput(value) {
      this.showEmailAutoComplete = value.includes('@')
    },
    completeEmail(domain) {
      const value = this.$attrs.value
      if (value.includes('@')) {
        const fullEmail =
          value.substring(0, value.lastIndexOf('@') + 1) + domain
        this.$emit('input', fullEmail)
      }
    },
  },
}
</script>

<style lang="scss" scoped>
::v-deep.v-text-field--outlined fieldset {
  border-color: var(--v-grey-lighten4);
  border-radius: 10px;
}
::v-deep {
  &.v-text-field--outlined.v-input--is-focused fieldset,
  &.v-text-field--outlined.v-input--has-state fieldset {
    border-width: 1px;
    border-color: currentColor;
  }
  .theme--light.v-label {
    color: var(--v-label-base);
  }

  .v-input__prepend-inner,
  .v-input__append-inner {
    margin-top: 0 !important;
    height: 100%;
    align-items: center;
  }
  .v-input__prepend-inner {
    padding-left: 3px;
    padding-right: 15px !important;
  }
  .v-input__append-inner {
    padding-right: 3px;
    padding-left: 15px !important;
  }
  .password-toggle {
    cursor: pointer;
    font-weight: normal;
    font-size: 13px;
    line-height: 16px;
    text-align: right;
    letter-spacing: -0.41px;
    color: var(--v-grey-lighten2);
  }
  .v-text-field__details {
    padding: 0 !important;
    margin-bottom: 0 !important;
  }
}
.auto-domains {
  display: flex;
  flex-wrap: wrap;
  margin-top: 8px;
  p {
    background: rgba(191, 213, 255, 0.4);
    border-radius: 10px;
    font-size: 11px;
    line-height: 14px;
    letter-spacing: -0.41px;
    padding: 5px 11px;
    margin-right: 6px;
    cursor: pointer;
    margin-bottom: 6px;
    @include breakpoint('xs') {
      margin-right: 2px;
      margin-bottom: 3px;
    }
  }
}
</style>
