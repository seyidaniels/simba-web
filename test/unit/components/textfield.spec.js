/* eslint-disable no-undef */
// eslint-disable-next-line no-unused-vars
import { mount, shallowMount } from '@vue/test-utils'
import AppTextField from '../../../components/Base/Forms/AppTextField'

describe('TextField Component', () => {
  const textField = mount(AppTextField)
  it('should show an input tag', function () {
    expect(textField.contains('input')).toBe(true)
  })
  it('shows an input tag with password field', () => {
    const passwordWrapper = mount(AppTextField, {
      propsData: {
        type: 'password',
      },
    })
    expect(passwordWrapper.html()).toMatchSnapshot()
  })
  it('expects icon to be appended', () => {
    const textFieldWrapper = mount(AppTextField, {
      props: {
        appendIcon: 'fa-user',
      },
    })
    expect(textFieldWrapper.html()).toMatchSnapshot()
  })

  it('expects icon to be pre-pended', () => {
    const prePendComponent = {
      components: {
        'app-text-field': AppTextField,
      },
      template: `
        <app-text-field
          placeholder="AppTextField with a custom prepend-inner icon"
        >
        <template #prepend-inner>
          <img
            width="24"
            height="24"
            src="https://cdns.iconmonstr.com/wp-content/assets/preview/2018/240/iconmonstr-random-thin.png"
          />
        </template>
        </app-text-field>
      `,
    }

    const textFieldWrapper = mount(prePendComponent)
    expect(textFieldWrapper.html()).toMatchSnapshot()
  })
})
