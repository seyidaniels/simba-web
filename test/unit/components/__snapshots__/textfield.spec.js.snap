// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Component expects icon to be appended 1`] = `
"<div class=\\"v-input v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined\\">
  <div class=\\"v-input__control\\">
    <div class=\\"v-input__slot\\" style=\\"height: 65px;\\">
      <fieldset aria-hidden=\\"true\\">
        <legend style=\\"width: 0px;\\"><span>​</span></legend>
      </fieldset>
      <div class=\\"v-text-field__slot\\"><input id=\\"input-8\\" type=\\"text\\"></div>
    </div>
  </div>
</div>"
`;

exports[`TextField Component expects icon to be pre-pended 1`] = `
"<div class=\\"v-input v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-text-field--placeholder\\" placeholder=\\"AppTextField with a custom prepend-inner icon\\">
  <div class=\\"v-input__control\\">
    <div class=\\"v-input__slot\\" style=\\"height: 65px;\\">
      <div class=\\"v-input__prepend-inner\\"><img width=\\"24\\" height=\\"24\\" src=\\"https://cdns.iconmonstr.com/wp-content/assets/preview/2018/240/iconmonstr-random-thin.png\\"></div>
      <fieldset aria-hidden=\\"true\\">
        <legend style=\\"width: 0px;\\"><span>​</span></legend>
      </fieldset>
      <div class=\\"v-text-field__slot\\"><input id=\\"input-6\\" placeholder=\\"AppTextField with a custom prepend-inner icon\\" type=\\"text\\"></div>
    </div>
  </div>
</div>"
`;

exports[`TextField Component shows an input tag with password field 1`] = `
"<div class=\\"v-input v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined\\" type=\\"password\\">
  <div class=\\"v-input__control\\">
    <div class=\\"v-input__slot\\" style=\\"height: 65px;\\">
      <fieldset aria-hidden=\\"true\\">
        <legend style=\\"width: 0px;\\"><span>​</span></legend>
      </fieldset>
      <div class=\\"v-text-field__slot\\"><input id=\\"input-5\\" type=\\"text\\"></div>
    </div>
  </div>
</div>"
`;
